worker/sshserver/server.go

```diff
@@ -0,0 +1,207 @@
+	// leave this value nil and a listener will be spawned.
+	Listener net.Listener
+}
+
+// Validate validates the workers configuration is as expected.
+func (c ServerWorkerConfig) Validate() error {
+	if c.Logger == nil {
+		return errors.NotValidf("Logger is required")
+	}
+	return nil
+}
+
+// ServerWorker is a worker that runs an ssh server.
+type ServerWorker struct {
+	tomb tomb.Tomb
+
+	// Server holds the embedded server.
+	Server *ssh.Server
+
+	// config holds the configuration required by the server worker.
+	config ServerWorkerConfig
+}
+
+// NewServerWorker returns a running embedded SSH server.
+func NewServerWorker(config ServerWorkerConfig) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	s := &ServerWorker{config: config}
+	s.Server = &ssh.Server{
+		Addr: ":2223",
+		PublicKeyHandler: func(ctx ssh.Context, key ssh.PublicKey) bool {
+			return true
+		},
+		PasswordHandler: func(ctx ssh.Context, password string) bool {
+			return true
+		},
+		ChannelHandlers: map[string]ssh.ChannelHandler{
+			"direct-tcpip": s.directTCPIPHandler,
+		},
+	}
+
+	// TODO(ale8k): Update later to use the host key from StateServingInfo
+	terminatingHostKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, errors.Annotate(err, "failed to generate host key")
+	}
+	signer, _ := gossh.NewSignerFromKey(terminatingHostKey)
+	if err != nil {
+		return nil, errors.Annotate(err, "failed to create signer")
+	}
+
+	s.Server.AddHostKey(signer)
+
+	s.tomb.Go(s.loop)
+
+	return s, nil
+}
+
+// Kill stops the server worker by killing the tomb. Implements worker.Worker.
+func (s *ServerWorker) Kill() {
+	s.tomb.Kill(nil)
+}
+
+// Wait waits for the server worker to stop. Implements worker.Worker.
+func (s *ServerWorker) Wait() error {
+	return s.tomb.Wait()
+}
+
+func (s *ServerWorker) loop() error {
+	go func() {
+		<-s.tomb.Dying()
+		if err := s.Server.Close(); err != nil {
+			// There's really not a lot we can do if the shutdown fails,
+			// either due to a timeout or another reason. So we simply log it.
+			s.config.Logger.Errorf("failed to shutdown server: %v", err)
+		}
+	}()
+
+	var err error
+	if s.config.Listener != nil {
+		err = s.Server.Serve(s.config.Listener)
+	} else {
+		err = s.Server.ListenAndServe()
+	}
+	if errors.Is(err, ssh.ErrServerClosed) {
+		return nil
+	}
+	return err
+
+}
+
+func (s *ServerWorker) directTCPIPHandler(srv *ssh.Server, conn *gossh.ServerConn, newChan gossh.NewChannel, ctx ssh.Context) {
+	d := struct {
+		DestAddr string
+		DestPort uint32
+		SrcAddr  string
+		SrcPort  uint32
+	}{}
+
+	if err := gossh.Unmarshal(newChan.ExtraData(), &d); err != nil {
+		err := newChan.Reject(gossh.ConnectionFailed, "Failed to parse channel data")
+		if err != nil {
+			s.config.Logger.Errorf("failed to reject channel: %v", err)
+		}
+		return
+	}
+
+	ch, reqs, err := newChan.Accept()
```

Comment:
These names are confusing, `ch` being a common name for a channel, which this is not.

Reply:
renamed to sshCh